Src/VsSpecific/VsSpecific/CSharpScriptExecutor.cs

```diff
@@ -0,0 +1,116 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+using System.Threading.Tasks;
+using Vim.Interpreter;
+
+namespace Vim.VisualStudio.Specific
+{
+    internal sealed class CSharpScriptExecutor : ICSharpScriptExecutor
+    {
+        private const string ScriptFolder = "vsvimscripts";
+        private Dictionary<string, Script<object>> _scripts = new Dictionary<string, Script<object>>(StringComparer.OrdinalIgnoreCase);
+        private ScriptOptions _scriptOptions = null;
+
+        private async Task ExecuteAsync(IVim vim, CallInfo callInfo, bool createEachTime)
+        {
+            try
+            {
+                Script<object> script;
+                if (!TryGetScript(vim, callInfo.Name, createEachTime, out script))
+                    return;
+
+                var globals = new CSharpScriptGlobals(callInfo, vim);
+                var scriptState = await script.RunAsync(globals);
+            }
+            catch (CompilationErrorException ex)
+            {
+                if (_scripts.ContainsKey(callInfo.Name))
+                    _scripts.Remove(callInfo.Name);
+
+                vim.ActiveStatusUtil.OnError(string.Join(Environment.NewLine, ex.Diagnostics));
+            }
+            catch (Exception ex)
+            {
+                vim.ActiveStatusUtil.OnError(ex.Message);
+            }
+        }
+
+        private static ScriptOptions GetScriptOptions(string scriptPath)
+        {
+            var ssr = ScriptSourceResolver.Default.WithBaseDirectory(scriptPath);
+            var baseDirectory = AppDomain.CurrentDomain.BaseDirectory;
+
+            var searchPaths = new string[]
+            {
+                Path.Combine(baseDirectory, "PublicAssemblies"),
+                Path.Combine(baseDirectory, "PrivateAssemblies"),
+                Path.Combine(baseDirectory, @"CommonExtensions\Microsoft\Editor"),
+                Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location)
+            };
+
+            var smr = ScriptMetadataResolver.Default
+                .WithBaseDirectory(scriptPath)
+                .WithSearchPaths(searchPaths);
+
+            var asm = new Assembly[]
+            {
+                typeof(Vim.IVim).Assembly,                        // VimCore.dll
+                typeof(Vim.UI.Wpf.IBlockCaret).Assembly,          // VimWpf.dll
+                typeof(Vim.VisualStudio.ISharedService).Assembly, // Vim.VisualStudio.VsInterfaces.dll
+                typeof(Vim.VisualStudio.Extensions).Assembly      // Vim.VisualStudio.Shared.dll
+            };
+
+            var so = ScriptOptions.Default
+                  .WithSourceResolver(ssr)
+                  .WithMetadataResolver(smr)
+                  .WithReferences(asm);
+
+            return so;
+        }
+
+        private bool TryGetScript(IVim vim, string scriptName, bool createEachTime, out Script<object> script)
+        {
+            if (!createEachTime && _scripts.ContainsKey(scriptName))

```

Comment:
Can be simplified to 

```
!createEachTime && _scripts.TryGetValue(scriptName, out script)
```

The version today uses double lookup with `Contains` and the indexer